[cpp/route_guide] Fix JSON minification removing spaces inside string values (#38954)

jiangqucheng opened the pull request:
[cpp/route_guide] Fix JSON minification removing spaces inside string values (#38954)

The existing JSON minification logic removes all whitespace characters using 'std::remove_if(db_.begin(), db_.end(), isspace);'. This approach inadvertently strips spaces inside JSON string values, leading to incorrect data parsing.

This commit introduces a new JSON minification function that:
- Removes spaces **outside** JSON string values.
- Preserves spaces **inside** JSON string values (`"..."`).
- Correctly handles escaped quotes (`\"`).

Fixes #38954




<!--

If you know who should review your pull request, please assign it to that
person, otherwise the pull request would get assigned randomly.

If your pull request is for a specific language, please add the appropriate
lang label.

-->
## Verification

### Before Fix
```bash
> ./cmake/build/route_guide_client
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1741496886.397564  341105 helper.cc:201] DB parsed, loaded 100 features.
-------------- GetFeature --------------
Found feature called BerkshireValleyManagementAreaTrail,Jefferson,NJ,USA at 40.9146, -74.6189
Found no feature at 0, 0
-------------- ListFeatures --------------
Looking for features between 40, -75 and 42, -73
Found feature called PatriotsPath,Mendham,NJ07945,USA at 40.7838, -74.6144
Found feature called 101NewJersey10,Whippany,NJ07981,USA at 40.8123, -74.3999
Found feature called U.S.6,Shohola,PA18458,USA at 41.3628, -74.9016
Found feature called 5ConnersRoad,Kingston,NY12401,USA at 42, -74.0371
Found feature called MidHudsonPsychiatricCenter,NewHampton,NY10958,USA at 41.4008, -74.3951
Found feature called 287FlugertownRoad,LivingstonManor,NY12758,USA at 41.9611, -74.6525
Found feature called 4001TremleyPointRoad,Linden,NJ07036,USA at 40.611, -74.2187
Found feature called 352SouthMountainRoad,Wallkill,NY12589,USA at 41.6802, -74.237
Found feature called BaileyTurnRoad,Harriman,NY10926,USA at 41.295, -74.1077
Found feature called 193-199WawayandaRoad,Hewitt,NJ07421,USA at 41.2145, -74.395
Found feature called 406-496WardAvenue,PineBush,NY12566,USA at 41.5737, -74.2848
...
```

### After Fix
```bash
> ./cmake/build/route_guide_client
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1741496644.493263  337350 helper.cc:203] DB parsed, loaded 100 features.
-------------- GetFeature --------------
Found feature called Berkshire Valley Management Area Trail, Jefferson, NJ, USA at 40.9146, -74.6189
Found no feature at 0, 0
-------------- ListFeatures --------------
Looking for features between 40, -75 and 42, -73
Found feature called Patriots Path, Mendham, NJ 07945, USA at 40.7838, -74.6144
Found feature called 101 New Jersey 10, Whippany, NJ 07981, USA at 40.8123, -74.3999
Found feature called U.S. 6, Shohola, PA 18458, USA at 41.3628, -74.9016
Found feature called 5 Conners Road, Kingston, NY 12401, USA at 42, -74.0371
Found feature called Mid Hudson Psychiatric Center, New Hampton, NY 10958, USA at 41.4008, -74.3951
Found feature called 287 Flugertown Road, Livingston Manor, NY 12758, USA at 41.9611, -74.6525
Found feature called 4001 Tremley Point Road, Linden, NJ 07036, USA at 40.611, -74.2187
Found feature called 352 South Mountain Road, Wallkill, NY 12589, USA at 41.6802, -74.237
Found feature called Bailey Turn Road, Harriman, NY 10926, USA at 41.295, -74.1077
Found feature called 193-199 Wawayanda Road, Hewitt, NJ 07421, USA at 41.2145, -74.395
Found feature called 406-496 Ward Avenue, Pine Bush, NY 12566, USA at 41.5737, -74.2848
...
```

jiangqucheng commented:
Sorry, I can't fix the `Mergeable — 1/1 Fail(s): LABEL` check, as I cannot edit the label in this repo. Can anyone tell me how to do that?

jiangqucheng commented:
Hello @yashykt , could you please help me go through this review process, thanks. If there is anything that need further clarification please comment below, I will try my best to solve

jiangqucheng commented:
Hello @yashykt , haven't heard feedback from you, could you please push forward on this pull request, Thanks so much. ⸜(｡˃ ᵕ ˂ )⸝♡

yashykt commented:
Thanks for the PR! I'll take a look

pawbhard commented on examples/cpp/route_guide/helper.cc:
```diff
@@ -134,12 +132,59 @@ class Parser {
-  db_content.erase(
-      std::remove_if(db_content.begin(), db_content.end(), isspace),
-      db_content.end());
+  std::string db_content(MinifyJson(db));
```
Thanks for raisng the PR. However we don't waht this complex logic to be part of our tests. The issue mentioned can be fixed by using the below regex.

```
    std::regex whitespaceOutsideQuotes(R"(\s+(?=(?:(?:[^"]*"){2})*[^"]*+$))");
    // Replace all matches with an empty string
    return std::regex_replace(db, whitespaceOutsideQuotes, "");
```

jiangqucheng replied:
> Function content can be pasted here with 2 line comments similar to before "// A simple parser for the json db file. It requires the db file to have the // exact form of [{"location":{"latitude":123,"longitude":456},"name": // "the name can be empty"},{ ... }..."

Done, comments added in commit [32a737](https://github.com/grpc/grpc/pull/38955/commits/32a737bf6521f9608113f2925e50d7f3cda51a14)

jiangqucheng commented:
Hello @pawbhard ,
Thanks for your great idea. 👍  I have corrected the regex expression, replaced the code in the last commit, and tested it in my local environment. Now it's off to go, I believe.
